### PR TITLE
♻️ Major cleanup of catalog service

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

This is a large cleanup refactoring of `catalog` before modifying its functionality for the related issues mentioned below.

### Highlights

- ♻️ Upgrades to using `Annotation`s
- ✨ Extends CLI with `echo-envfile`
- 🗑️ Removes deprecated local `env-devel` and `extra-docker-compose.yaml` 
   - adapts `Makefile` and service CLI with `echo-envfile` command
   - simplifies `README`
- ♻️ `services/catalog/src/simcore_service_catalog/core/background_tasks.py`
   -  move implementation to services
   - make sure only one replica runs it
- ♻️ separated exception classes and handlers
- ♻️ test **fixtures**: 
       - enhanced and reduced tests fixtures by using `pytest_simcore` assets
- ♻️ ruffed entire code base

## Related issue/s

- Follows up from https://github.com/ITISFoundation/osparc-simcore/pull/5902
- Preparation for https://github.com/ITISFoundation/osparc-issues/issues/1404

## How to test

```cmd
cd services/catalog
make install-dev
make test-dev-unit
```

## Dev-ops checklist

- `CATALOG_BACKGROUND_TASK_REST_TIME` was missing
- Please review https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/merge_requests/622

### CODE REVIEW

File: services/catalog/src/simcore_service_catalog/core/events.py [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
 from .background_tasks import start_registry_sync_task, stop_registry_sync_task
 
-logger = logging.getLogger(__name__)
+_logger = logging.getLogger(__name__)
 
 
-def create_start_app_handler(app: FastAPI) -> Callable:
-    async def start_app() -> None:
-        logger.info("Application started")
+def create_on_startup(app: FastAPI):
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/core/events.py:14]:
```suggestion
def create_on_startup(app: FastAPI) -> Callable[[], Awaitable[None]]:
```

File: services/catalog/src/simcore_service_catalog/core/events.py [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
+    async def _():
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/core/events.py:15]:
```suggestion
    async def _() -> None:
```

File: services/catalog/src/simcore_service_catalog/core/events.py [hunk 3]
```diff
@@ -28,20 +28,23 @@ async def start_app() -> None:
             # SEE https://github.com/ITISFoundation/osparc-simcore/issues/1728
             await start_registry_sync_task(app)
 
-    return start_app
+        _logger.info("Application started")
 
+    return _
+
+
+def create_on_shutdown(app: FastAPI):
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/core/events.py:36]: similarly here and in the callable inside

File: services/catalog/src/simcore_service_catalog/db/repositories/services.py [hunk 4]
```diff
@@ -66,6 +66,28 @@ def _make_list_services_query(
     return query
 
 
+def _is_newer(
+    old: ServiceSpecificationsAtDB | None,
+    new: ServiceSpecificationsAtDB,
+):
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/db/repositories/services.py:72]:
```suggestion
) -> bool:
```

File: services/catalog/src/simcore_service_catalog/exceptions/handlers/__init__.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+from fastapi import FastAPI, HTTPException, status
+from fastapi.exceptions import RequestValidationError
+
+from ._http_error import http_error_handler, make_http_error_handler_for_exception
+from ._validation_error import http422_error_handler
+
+
+def setup_exception_handlers(app: FastAPI):
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/exceptions/handlers/__init__.py:8]:
```suggestion
def setup_exception_handlers(app: FastAPI) -> None:
```

File: services/catalog/tests/unit/with_dbs/test_api_routes_services__list.py [hunk 6]
```diff
@@ -27,21 +29,20 @@
 
 
 @pytest.fixture
-def disable_service_caching(monkeypatch):
-    monkeypatch.setenv("AIOCACHE_DISABLE", 1)
+def disable_service_caching(monkeypatch: pytest.MonkeyPatch):
```
Comment (GitHK) [inline services/catalog/tests/unit/with_dbs/test_api_routes_services__list.py:32]:
```suggestion
def disable_service_caching(monkeypatch: pytest.MonkeyPatch) -> None:
```

File: services/catalog/src/simcore_service_catalog/cli.py [hunk 7]
```diff
@@ -24,3 +27,52 @@ def run():
         "$ uvicorn simcore_service_catalog.main:the_app",
         fg=typer.colors.BLUE,
     )
+
+
+@main.command()
+def echo_dotenv(ctx: typer.Context, *, minimal: bool = True):
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/cli.py:33]:
```suggestion
def echo_dotenv(ctx: typer.Context, *, minimal: bool = True) -> None:
```

**Review by GitHK — APPROVED** [review]
👍

File: services/catalog/requirements/_test.txt [hunk 8]
```diff
@@ -1,180 +1,115 @@
...
 certifi==2024.2.2
     # via
-    #   -c requirements/../../../requirements/constraints.txt
-    #   -c requirements/_base.txt
     #   httpcore
     #   httpx
     #   requests
 charset-normalizer==3.3.2
     # via requests
 click==8.1.7
```
Comment (sanderegg) [inline services/catalog/requirements/_test.txt:24]: aren't you using an old uv here? why are all the via going away?

File: services/catalog/src/simcore_service_catalog/api/routes/services.py [hunk 9]
```diff
@@ -76,19 +76,23 @@ def _build_cache_key(fct, *_, **kwargs):
 # (when e2e runs or by the webserver when listing projects) therefore
 # a cache is setup here
 @router.get("", response_model=list[ServiceGet], **RESPONSE_MODEL_POLICY)
-@cancellable_request
 @cached(
     ttl=LIST_SERVICES_CACHING_TTL,
     key_builder=_build_cache_key,
 )
+@cancel_on_disconnect
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/api/routes/services.py:83]: Q: why did you move that one? just wondering if that makes sense. caching a cancellable request? should the cancellation not be on top?

File: services/catalog/src/simcore_service_catalog/core/background_tasks.py [hunk 10]
```diff
@@ -29,12 +29,14 @@
...
 from ..services import access_rights
 
-logger = logging.getLogger(__name__)
+_logger = logging.getLogger(__name__)
 
 # NOTE: by PC I tried to unify with models_library.services but there are other inconsistencies so I leave if for another time!
-ServiceKey = str
-ServiceVersion = str
-ServiceDockerDataMap = dict[tuple[ServiceKey, ServiceVersion], ServiceDockerData]
+ServiceKey: TypeAlias = str
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/core/background_tasks.py:35]: we still need to discuss ```NewType``` vs ```TypeAlias```

File: services/catalog/src/simcore_service_catalog/db/repositories/projects.py [hunk 11]
```diff
@@ -17,7 +17,7 @@ async def list_services_from_published_templates(self) -> list[ServiceKeyVersion
             async for row in await conn.stream(
                 sa.select(projects).where(
                     (projects.c.type == ProjectType.TEMPLATE)
-                    & (projects.c.published == True)
+                    & (projects.c.published is True)
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/db/repositories/projects.py:20]: is this legal? should you not use ```.is_(True)``` ??

File: services/catalog/src/simcore_service_catalog/services/access_rights.py [hunk 12]
```diff
@@ -106,7 +106,7 @@ async def evaluate_auto_upgrade_policy(
     # AUTO-UPGRADE PATCH policy:
     #
     #  - Any new patch released, inherits the access rights from previous compatible version
-    #  - TODO: add as option in the publication contract, i.e. in ServiceDockerData
+    #  - IDEA: add as option in the publication contract, i.e. in ServiceDockerData?
```
Comment (sanderegg) [inline services/catalog/src/simcore_service_catalog/services/access_rights.py:109]: mouarf... ;)

**Review by sanderegg — COMMENTED** [review]
please check the cancellation decorator. I find weird to place it after the cached one

File: services/catalog/src/simcore_service_catalog/core/background_tasks.py (hunk 10 above)
Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/core/background_tasks.py:35]: Ah I see your point. `NewType("ServiceKey", str)` would be something like
```python
class NewType(str):
     ...
```

And therefore creating a new type as a subclass of `str`.

File: services/catalog/src/simcore_service_catalog/api/routes/services.py (hunk 9 above)
Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/api/routes/services.py:83]: yes, my bad. good catch. I think it is safer to put it on top

File: services/catalog/src/simcore_service_catalog/db/repositories/projects.py (hunk 11 above)
Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/db/repositories/projects.py:20]: Goot catch ... ! I believe it was rust changes ... but not sure.
The question here is why the tests did not catch that! Will also increase coverage a bit more before start doing changes

Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/db/repositories/projects.py:20]: @sanderegg FYI:
```cmd
>>> from simcore_postgres_database.models.projects import projects
>>> from sqlalchemy.dialects import postgresql
>>> import sqlalchemy as sa
```
then 
```cmd
>>> stmt = sa.select(projects).where(projects.c.published is True)
>>> print(stmt.compile(dialect=postgresql.dialect(), compile_kwargs={"literal_binds": True}))
SELECT projects.id, projects.type, projects.uuid, projects.name, projects.description, projects.thumbnail, projects.prj_owner, projects.creation_date, projects.last_change_date, projects.access_rights, projects.workbench, projects.ui, projects.classifiers, projects.dev, projects.quality, projects.published, projects.hidden 
FROM projects 
WHERE false
```
vs
```
>>> stmt = sa.select(projects).where(projects.c.published.is_(True))
>>> print(stmt.compile(dialect=postgresql.dialect(), compile_kwargs={"literal_binds": True}))
SELECT projects.id, projects.type, projects.uuid, projects.name, projects.description, projects.thumbnail, projects.prj_owner, projects.creation_date, projects.last_change_date, projects.access_rights, projects.workbench, projects.ui, projects.classifiers, projects.dev, projects.quality, projects.published, projects.hidden 
FROM projects 
WHERE projects.published IS true
```

File: services/catalog/src/simcore_service_catalog/services/access_rights.py (hunk 12 above)
Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/services/access_rights.py:109]: it is that an expression of love or discust :-D

File: services/catalog/requirements/_test.txt (hunk 8 above)
Comment (pcrespov) [reply services/catalog/requirements/_test.txt]: yes i was. thx!

File: .env-devel [hunk 13]
```diff
@@ -30,6 +30,7 @@ AUTOSCALING_LOGLEVEL=WARNING
 AUTOSCALING_NODES_MONITORING=null
 AUTOSCALING_POLL_INTERVAL=10
 
+CATALOG_BACKGROUND_TASK_REST_TIME=60
```
Comment (YuryHrytsuk) [inline .env-devel:33]: I cannot find this env var in `docker-compose.yml`. Is it on purpose?

**Review by YuryHrytsuk — APPROVED** [review]

File: .env-devel (hunk 13 above)
Comment (pcrespov) [reply .env-devel:33]: It was already there. The new testing detected that it was not in `.env-devel`
image

File: services/catalog/src/simcore_service_catalog/api/routes/services_resources.py [hunk 14]
```diff
@@ -150,7 +150,7 @@ async def _get_service_labels(
         # and will fail validating the key or the version
         if err.status_code == status.HTTP_400_BAD_REQUEST:
             return None
-        raise err
+        raise
```
Comment (bisgaard-itis) [inline services/catalog/src/simcore_service_catalog/api/routes/services_resources.py:153]: For my learning: What's actually the difference between `raise err` and `raise` here?

File: services/catalog/src/simcore_service_catalog/cli.py [hunk 15]
```diff
@@ -24,3 +27,52 @@ def run():
         "$ uvicorn simcore_service_catalog.main:the_app",
         fg=typer.colors.BLUE,
     )
+
+
+@main.command()
+def echo_dotenv(ctx: typer.Context, *, minimal: bool = True) -> None:
```
Comment (bisgaard-itis) [inline services/catalog/src/simcore_service_catalog/cli.py:33]: maybe just `echo_env`? 🤔

File: services/catalog/src/simcore_service_catalog/cli.py [hunk 16]
```diff
@@ -24,3 +27,52 @@ def run():
...
+    assert ctx  # nosec
+
+    # NOTE: we normally DO NOT USE `os.environ` to capture env vars but this is a special case
+    # The idea here is to have a command that can generate a **valid** `.env` file that can be used
+    # to initialized the app. For that reason we fill required fields of the `ApplicationSettings` with
+    # "fake" but valid values (e.g. generating a password or adding tags as `replace-with-api-key).
+    # Nonetheless, if the caller of this CLI has already some **valid** env vars in the environment we want to use them ...
+    # and that is why we use `os.environ`.
+
+    settings = ApplicationSettings.create_from_envs(
```
Comment (bisgaard-itis) [inline services/catalog/src/simcore_service_catalog/cli.py:50]: I know that we often have example of valid data for the pydantic models in the `config` in the models. These are often tested, so that we know the data is valid. Maybe it could be make sense to use that here.

**Review by bisgaard-itis — APPROVED** [review]
Very cool! Thanks a lot. 
Just left a few suggestions

File: services/catalog/src/simcore_service_catalog/cli.py (hunk 15 above)
Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/cli.py:33]: the files are commonly denoted as dot-env or `.env` :-) ... but it is a matter of taste.  In any case, the important thing is to use the same command across all the service settings

File: services/catalog/src/simcore_service_catalog/api/routes/services_resources.py (hunk 14 above)
Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/api/routes/services_resources.py:153]: In this context it is the same. 
`raise` alone is a short way of saying "re-raise the exception caught in this context"

File: services/catalog/Makefile [hunk 17]
```diff
@@ -5,69 +5,19 @@ include ../../scripts/common.Makefile
...
 .PHONY: openapi-specs openapi.json
 openapi-specs: openapi.json
-openapi.json: .env
-	# generating openapi specs file
+openapi.json: .env-ignore
+	# generating openapi specs file (need to have the environment set for this)
+	@set -o allexport; \
+	source $<; \
+	set +o allexport; \
 	python3 -c "import json; from $(APP_PACKAGE_NAME).main import *; print( json.dumps(the_app.openapi(), indent=2) )" > $@
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/catalog/Makefile:20]: Many of these can be indeed be factored out in `scripts/*.Makefile`. I would do it for the entire repo and a separate PR

File: scripts/common-service.Makefile [hunk 18]
```diff
@@ -180,3 +180,23 @@ _run-test-ci: _check_venv_active
...
+OPENAPI_GENERATOR_NAME := itisfoundation/openapi-generator-cli-openapi-generator-v4.2.3
+OPENAPI_GENERATOR_TAG := v0
+OPENAPI_GENERATOR_IMAGE := $(OPENAPI_GENERATOR_NAME):$(OPENAPI_GENERATOR_TAG)
+
+define validate_openapi_specs
+	# Validating OAS '$(1)' ...
+	docker run --rm \
+			--volume "$(CURDIR):/local" \
+			$(OPENAPI_GENERATOR_IMAGE) validate --input-spec /local/$(strip $(1))
+endef
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply scripts/common-service.Makefile:202]: yes. We can also factor out as I mentione below

File: services/catalog/src/simcore_service_catalog/cli.py (hunk 16 above)
Comment (pcrespov) [reply services/catalog/src/simcore_service_catalog/cli.py:50]: @bisgaard-itis In reality the idea here is that provided some env vars in your system ... you are able to "extract" the minimum dotenv file that you need to run your service.   If some of the env vars are not present, we fill them with some "dummy" value

**Comment by matusdrobuliak66** [comment]
Thanks :+1:
